Feat(#31): Add ignore option

hinok opened the pull request:
- This commit add functionality to ignore files
based on provided regexps.
- in index.js config must be specified as below

    // @create-index {"ignore":["/foo.js"]}

gajus commented on src/utilities/log.js:
```diff
@@ -2,8 +2,6 @@ import chalk from 'chalk';
 
 export default (...append) => {
-    /* eslint-disable no-console */
+  /* eslint-disable no-console */
```
Use `// eslint-disable-next-line`

gajus commented on src/utilities/constants.js:
```diff
@@ -0,0 +1,3 @@
+const CREATE_INDEX_PATTERN = /(?:^|[\n\r]+)\/\/ @create-index\s?({.*})?[\n\r]+/;
```
`export const CREATE_INDEX_PATTERN ...`

gajus commented on package.json:
```diff
@@ -48,7 +48,8 @@
     "precommit": "npm run test",
-    "test": "npm run build && npm run lint && cross-env NODE_ENV=development mocha --compilers js:babel-register"
+    "test": "npm run build && npm run lint && cross-env NODE_ENV=development mocha --compilers js:babel-register",
+    "test:watch": "cross-env NODE_ENV=development mocha --watch --reporter spec --bail --compilers js:babel-register"
```
Remove `test:watch`.

gajus commented:
Thank you for the PR.

This lack documentation, `README.md` changes.

gajus commented on test/writeIndex.js:
```diff
@@ -8,29 +8,59 @@ import {
-    writeIndex([path.resolve(fixturesPath, 'mixed')]);
+  it('creates index with config in target directory', () => {
+    const indexFilePath = path.resolve(fixturesPath, 'with-config/index.js');
+    const ignoredExportLine = `export { default as bar } from './bar.js';`; // eslint-disable-line quotes
```
Use `// eslint-disable-next-line`

gajus commented on test/writeIndex.js:
```diff
@@ -8,29 +8,59 @@ import {
+};
+
+const appendToFile = (filePath, content) => {
+  try {
```
Whats the reason for swallowing an error?

gajus commented on test/writeIndex.js:
```diff
@@ -8,29 +8,59 @@ import {
+
+const removeFileByPath = (filePath) => {
+  // eslint-disable-next-line no-empty
+  try {
```
Whats the reason for swallowing an error?

hinok replied:
I was trying to follow "code style" in the project so if you look here https://github.com/gajus/create-index/pull/37/files#diff-f546cc5537a07f2741198b32e7ef3aa9L17 the same pattern with swallowing an error has been already used. I decided to just copy and use it.

Anyway I will remove them if you want, no problem :)

gajus replied:
Oh, thats in the `test/`. Meh, I am indifferent.

gajus commented on src/utilities/readIndexConfig.js:
```diff
@@ -0,0 +1,40 @@
+import hasIndex from './hasIndex';
+import {CREATE_INDEX_PATTERN} from './constants';
+
+const isString = (str) => {
```
Remove unnecessary helpers. I am all pro declarative programming, though `typeof str === 'string'` is already achieving the same.

gajus commented on src/utilities/readIndexConfig.js:
```diff
@@ -0,0 +1,40 @@
+};
+
+const isEmptyString = (str) => {
+  return str.length === 0;
```
Remove unnecessary helpers. I am all pro declarative programming, though `str.length` is already achieving the same.

hinok commented:
Sorry for so many changes. I applied all requested changes and removed "swallowing errors" from tests. I added also section about `ignore` in `README.md`.

gajus commented:
Is this ready to be merged?

hinok commented:
I thought that `writeIndexCli.js` uses `writeIndex.js`, it doesn't - my fault, I didn't check but I fixed it already.

I also checked this PR on my local project where I use `create-index` heavily to maintain icons, it works 💃 

Please feel free to correct any english language issues.
I think that it's ready to be merged.

gajus commented:
Thank you
